ray-operator/controllers/ray/raycluster_controller_test.go

```diff
@@ -225,6 +225,96 @@ var _ = Context("Inside the default namespace", func() {
 		})
 	})
 
+	Describe("RayCluster with overridden app.kubernetes.io labels", func() {
+		ctx := context.Background()
+		namespace := "default"
+		rayCluster := rayClusterTemplate("raycluster-overridden-k8s-labels", namespace)
+		rayCluster.Spec.HeadGroupSpec.Template.Labels = map[string]string{
+			utils.KubernetesApplicationNameLabelKey: "myapp",
+		}
+		headPods := corev1.PodList{}
+		workerPods := corev1.PodList{}
+		workerFilters := common.RayClusterGroupPodsAssociationOptions(rayCluster, rayCluster.Spec.WorkerGroupSpecs[0].GroupName).ToListOptions()
+		headFilters := common.RayClusterHeadPodsAssociationOptions(rayCluster).ToListOptions()
+
+		It("Verify RayCluster spec", func() {
+			// These test are designed based on the following assumptions:
+			// (1) The app.kubernetes.io/name label of the HeadGroupSpec is overridden.
+			// (2) There is only one worker group, and its `replicas` is set to 3.
+			Expect(rayCluster.Spec.HeadGroupSpec.Template.Labels[utils.KubernetesApplicationNameLabelKey]).NotTo(BeEmpty())
+			Expect(len(rayCluster.Spec.WorkerGroupSpecs)).To(Equal(1))
+			Expect(rayCluster.Spec.WorkerGroupSpecs[0].Replicas).To(Equal(pointer.Int32(3)))
+		})
+
+		It("Create a RayCluster custom resource", func() {
+			err := k8sClient.Create(ctx, rayCluster)
+			Expect(err).NotTo(HaveOccurred(), "Failed to create RayCluster")
+			Eventually(
+				getResourceFunc(ctx, client.ObjectKey{Name: rayCluster.Name, Namespace: namespace}, rayCluster),
+				time.Second*3, time.Millisecond*500).Should(BeNil(), "Should be able to see RayCluster: %v", rayCluster.Name)
+		})
+
+		It("Check the number of head Pods", func() {
+			numHeadPods := 1
+			Eventually(
+				listResourceFunc(ctx, &headPods, headFilters...),
+				time.Second*3, time.Millisecond*500).Should(Equal(numHeadPods), fmt.Sprintf("headGroup %v", headPods.Items))
+			for _, head := range headPods.Items {
+				Expect(head.Labels[utils.KubernetesApplicationNameLabelKey]).To(Equal("myapp"))
+			}
+		})
+
+		It("Check the number of worker Pods", func() {
+			numWorkerPods := 3
+			Eventually(
+				listResourceFunc(ctx, &workerPods, workerFilters...),
+				time.Second*3, time.Millisecond*500).Should(Equal(numWorkerPods), fmt.Sprintf("workerGroup %v", workerPods.Items))
+		})
+
+		It("Update all Pods to Running", func() {
```

Comment:
Do we need to update all pods to "running"? Can we just check whether `RayCluster.Status.Head.ServiceIP` is set correctly? I am not sure whether `Service.Spec.ClusterIP` will be set automatically or not in the envtest.

Reply:
Unfortunately, IP addresses like `pod.Status.PodIP` will not be populated in envtests. And we use Headless services by default now. The value of `Service.Spec.ClusterIP` will be `None`.

Checking `RayCluster.Status.Head.ServiceIP` looks good. That makes the test more explicit. I have added it.

However, I think we should still update all pods to "running" and check if the RayCluster will be ready for two reasons:
1. Checking `RayCluster.Status.Head.ServiceIP` solely makes the test incomplete and weird: "First override the app.kubernetes.io labels, and then end the test by checking the ServiceIP". It is hard for others to understand why the test is written this way if they don't inspect the whole commit.
2. By making sure the RayCluster will be ready, we can avoid future changes that carry unexpected labels into other resource selectors.